modeltranslation/fields.py

```diff
@@ -256,6 +258,17 @@ def __ne__(self, other):
     def __hash__(self):
         return hash((self.creation_counter, self.language))
 
+    def get_default(self):
+        with override(self.language):
+            default = super().get_default()
+            # we must *force evaluation* at this point, otherwise the lazy translatable
+            # string is returned and will be evaluated only later when the main language
+            # is activated again (because this context block has exited).
+            #
+            # force_str passes protected types as-is, which includes None, int, float,
+            # datetime...
+            return force_str(default, strings_only=True)
```

Comment:
@last-partizan @bart-maykin took me some serious digging to figure out why the `gettext_lazy("foo")` variant didn't work as expected, but now I finally understand.

During `Model.__init__`, django populates the field defaults using `field.get_default()` if no value was provided (nothing ordinary here yet). The patch with `with override(self.language):` was correctly working and printing the default in the shell also produced the expected translated string. However, when actually running the tests without breakpoints, only the German translation were every reported.

This is because the project language is set to German and the lazy translated string was being returned all the way - effectively the `translation.override` was not having any effect. The lazy string is returned from `get_default`, and only when values are effectively evaluated (e.g. when doing `instance.foo == "bar"` assertions or when it's being written to the database) the lazy string is evaluated, but it's now using `settings.LANGUAGE_CODE` because the context block was exited.

As far as I can tell the only way to prevent this is by forcing the evaluation sooner and still inside the context manager, using the `force_str` utility. This one has a option where it passes through anything that's not like a string (so it would be safe to use this default with non-string based fields or `None` default values as employed by this library!). Locally the rest of the test suite is passing, which confirms this.

I don't think there are any negative drawbacks to this approach, but of course I could be missing something.

Reply:
Yes, that's fine!

Reply:
Looking at this again, and i noticed migrations are using english version for `dynamic_default_de` field.

Is that okay for now? Or maybe we can move this language-dependent evaluation somewhere where it would work for both runtime code and migrations?

Reply:
I believe this is not something this library can fix because (need to confirm this) django effectively disables real translations during makemigrations. That's also why the english string is baked in everywhere  rather than applying `settings.LANGUAGE_CODE` - it's taking the literal from the code.

If you really need dynamic values to be baked in the migration itself, you must use a callable default and not just a lazy value.

If django didn't do this, it would generate new migrations in your site-packages for any library using gettext-lazy in their model field definitions as soon as your project uses a different language than what they were generated with, which is of course unacceptable.

Probably good to mention this in the documentation?

Reply:
Yes, explaining this in the docs is great idea.

Reply:
But, before doing that - take a look at `TranslationField.deconstruct`. I think we can handle this problem there.

Reply:
I don't think so, because the [management command](https://github.com/django/django/blob/cafe7266ee69f7e017ddbc0d440084ace559b04b/django/core/management/commands/makemigrations.py#L105) is decorated with `@no_translations` which even deactivates all translations even if `USE_I18N = True` in the settings.

I suppose that in theory you could re-activate and re-deactivate translations in the `deconstruct`, but it seems to diverge pretty strongly from expected django behaviour. I'm not sure how invasive you want this library to be on this aspect, as the functional use case makes a lot of sense to me indeed.

Reply:
Ok, let's ignore this problem with migrations. If people want translated defaults here, they can always add it manyally.